src/guest_memory.rs

```diff
@@ -646,15 +654,15 @@ pub trait GuestMemory {
                 Ok(0) => return Ok(total),
                 // made some progress
                 Ok(len) => {
-                    total += len;
-                    if total == count {
-                        break;
-                    }
+                    total = match total.checked_add(len) {
+                        Some(x) if x < count => x,
+                        Some(x) if x == count => return Ok(x),
+                        _ => return Err(Error::CallbackOutOfRange),
+                    };
                     cur = match cur.overflowing_add(len as GuestUsize) {
-                        (GuestAddress(0), _) => GuestAddress(0),
-                        (result, false) => result,
-                        (_, true) => panic!("guest address overflow"),
-                    }
+                        (x @ GuestAddress(0), _) | (x, false) => x,
+                        (_, true) => return Err(Error::GuestAddressOverflow),
+                    };
```

Comment:
I think the `overflowing_add` would probably better be a `checked_add` too, but I guess that can be fixed separately.

Reply:
What is the reason for using `overflowing_add` instead of `checked_add`? And why would we return x if it overflows to 0. Wouldn't that also be an invalid access?

Reply:
I do not know, I have not investigated all the reasoning behind this. For now I'm preserving previous behavior, I think further improvements can be made outside of this PR.

Reply:
Oh, I didn't see that's how it was before. This should indeed be fixed in a follow-up PR.